nadabot/models.py

```diff
@@ -0,0 +1,234 @@
+        verbose_name=_("submitted by"),
+        max_length=100,
+        null=False,
+        help_text=_("User who submitted this provider.")
+    )
+    submitted_at_ms = models.DateTimeField(
+        _("submitted at (milliseconds)"),
+        null=False,
+        help_text=_("Timestamp of when this provider was submitted.")
+    )
+    stamp_validity_ms = models.BigIntegerField(
+        _("stamp validity"),
+        blank=True,
+        null=True,
+        help_text=_("Milliseconds that stamps from this provider are valid for before they expire.")
+    )
+    account_id_arg_name = models.CharField(
+        _("account ID argument name"),
+        max_length=100,
+        null=False,
+        help_text=_("Name of account ID argument, e.g. 'account_id' or 'accountId' or 'account'.")
+    )
+    custom_args = models.CharField(
+        _("custom args"),
+        null=True,
+        help_text=_("Custom args as Base64VecU8.")
+    )
+    registry = models.ForeignKey(
+        NadabotRegistry,
+        on_delete=models.CASCADE,
+        verbose_name=_("registry"),
+        help_text=_("registry under which provider was registered")
+    )
+
+    class Meta:
+        indexes = [models.Index(fields=["id", "status"], name="idx_provider_id_status")]
+
+
+class Stamp(models.Model):
+    user = models.ForeignKey(
+        Account,
+        on_delete=models.CASCADE,
+        verbose_name=_("user"),
+        help_text=_("The user who earned the stamp.")
+    )
+    provider = models.ForeignKey(
+        Provider,
+        on_delete=models.CASCADE,
+        verbose_name=_("provider"),
+        help_text=_("The provider the user verified with.")
+    )
+    verification_date = models.DateField(
+        _("verification date"),
+        auto_now_add=True,
+        help_text=_("The date of verification.")
+    )
+
+class Rule(models.Model):
+    id = models.AutoField(
+        _("rule id"),
+        primary_key=True,
+        help_text=_("rule id for group rules"),
+    )
+    rule = models.CharField(
+        _("rule enum choice"),
+        max_length=50,
+        null=False,
+        help_text=_("particular rule choice from the rule enum")
+    )
+    value = models.PositiveIntegerField(
+        _("rule value"),
+        null=True,
+        help_text=_("An optional value that rule choices might carry."),
+    )
+
+class Group(models.Model):
+    id = models.PositiveIntegerField(
+        _("group id"),
+        primary_key=True,
+        help_text=_("Group id."),
+    )
+    name = models.CharField(
+        _("group name"),
+        max_length=100,
+        null=False,
+        help_text=_("name given to the group")
+    )
+    group_rule = models.ForeignKey(
+        Rule,
+        on_delete=models.CASCADE,
+        verbose_name=_("rule"),
+        help_text=_("The rule this group uses.")
+    )
+    providers = models.ManyToManyField(
+        Provider,
+        related_name="group_provider",
+        help_text=_("group providers."),
+    )
```

Comment:
The more I think about it, the more having `Rule` as its own model feels weird to me, since it only ever exists within the context of a Group. Also, rules will be inconsistently shared between providers; e.g. all `Highest` and `Lowest` groups will use the same rule (respectively), but other rules will likely be unique records since they contain optional values.

How about just adding them as fields on Group instead, e.g.

`Group.rule_type`
`Group.rule_value`